### ISSUE TITLE
Socket connection fails in background with io error or No SRV result

### ISSUE THREAD

**Issue by JitainSharma** [BelledonneCommunications/linphone-android#1069]
Platform: Android
OS: Android 10
API: 29
Device: Samsung Galaxy M21
Internet Connection: WIFI
linphone-sdk: 4.3.3

While the app is in foreground or background, continuous socket error is coming due to which app is unable to connect with the server and calls fail.

This is happening at a higher rate when the app is in doze mode and connected to the wifi or disconnected and connected again to the wifi while in doze mode [Device is locked and the screen is off].

This issue is more prominent with the Android 10 devices.

I am attaching the logs file for the reference and seek help to resolve this issue. I am using the onSip server.

To point at right log, please have a look at the following logs:
- DNS resolution failed for edge.sip.onsip.com
- No SRV result for [_sip._tcp.edge.sip.onsip.com], trying A/AAAA
- [liblinphone] MESSAGE Register refresher [503] reason [io error] for proxy [<sip:edge.sip.onsip.com;transport=tcp>]

[linphone4.log](https://github.com/BelledonneCommunications/linphone-android/files/4812694/linphone4.log)

[linphone3.log](https://github.com/BelledonneCommunications/linphone-android/files/4812714/linphone3.log)

[linphone1_redmi.log](https://github.com/BelledonneCommunications/linphone-android/files/4812737/linphone1_redmi.log)

**Comment by Viish** [comment]
Hi,
An app can't do network operations while in doze mode.
However when a push is received you have a window of 20 seconds (it's an estimate and can vary depending on the device) to do your requests.
From your logs when the push is received the network operation is OK, otherwise it isn't so everything behaves like expected.
Cheers,

**Comment by JitainSharma** [comment]
Hi,

Thanks for pointing out the doze mode, we have already handled these in the app.

The problem is with the background running of the SDK, itself it makes the connection and fails, see the logs.

In the background, SDK listens to the doze mode, when it brings out of it and checks the network connectivity using: **ConnectivityManager getAllNetworkInfo NetworkInfo isConnectedOrConnecting** which tells on device-level, not on an application level.

Also, from Android 10, we can't start service from the background, by which the service **LinphoneService** fails to start.

Also, the problem is we're using a job scheduled via WorkManager for the registering on register push received, the core always prints the log as **Refresh register operation not available (network unreachable)** while the network is available.

Please help to resolve these issues or the workaround to resolve them.

**Comment by Viish** [comment]
Indeed there is another mode in which the device isn't allowed to make network operations, but as the doze mode it is disabled for a while once you receive a push.

You can start a service from the push notification handler (as a foreground service). If you rely on push notifications you can also let your app die when in background and the push will always create a new instance of your app. As long as the push as a "visible" effect (like a notification) it will be ok. You can check how I reworked all that in the feature/kotlin_rewrite branch. We use our app daily on Android 10 and it works great no matter the device state.

Anyway I believe that using a WorkManager isn't the right solution and also that it is the reason you can't do network operations even after receiving a push.

**Comment by JitainSharma** [comment]
Thanks @Viish for the details explanation, will go through the feature/kotlin_rewrite branch and also the suggestions you have described.

**Comment by JitainSharma** [comment]
Hi, just quick update, while in no background mode running and INVITE push is received, Incoming call is opened first, and after **ConnectivityManager** is registered to check for the internet connectivity where `onCapabilitiesChanged` is called and we register the core, after it we call `core.ensureRegistered()`.

But the registration is always failed with following code and INVITE message never received, this happens continiously.
I am pasting the snippet of that log below and full logs file. Please see the logs for date 2020-07-22.

My prediction, this is happening when phone goes into doze mode for more than 3-4 hrs.

> 2020-07-22 10:36:30:622 [liblinphone] MESSAGE _linphone_update_dependent_proxy_config(): 0x7d26710400 is registered, checking for [0x7d26710400] ->dependency=0x0
> 2020-07-22 10:36:30:624 [liblinphone] MESSAGE Linphone core [0x7d28c73600] notified [registration_state_changed]
> 2020-07-22 10:36:30:624 [belle-sip] MESSAGE channel [0x7c9858c6c0]: ending send background task with id=[4412].
> 2020-07-22 10:36:30:625 [belle-sip] MESSAGE bellesip_wake_lock_release(): Android wake lock released [ref=0x4412]
> 2020-07-22 10:36:30:625 [belle-sip] MESSAGE transaction [0x7d28c59900] channel state changed to [ERROR]
> 2020-07-22 10:36:30:625 [belle-sip] MESSAGE Refresher[0x7d28e1b900]: scheduling next timer in 60000 ms for purpose [retry]
> 2020-07-22 10:36:30:625 [liblinphone] MESSAGE Register refresher [503] reason [io error] for proxy [<sip:edge.sip.onsip.com;transport=tcp>]
> 2020-07-22 10:36:30:626 [liblinphone] MESSAGE Proxy config [0x7d26710400] for identity [sip:rs_2@sip1.mirconsip.com] moving from state [LinphoneRegistrationProgress] to [LinphoneRegistrationFailed] on core [0x7d28c73600]

[linphone1.txt](https://github.com/BelledonneCommunications/linphone-android/files/4958482/linphone1.txt)

**Comment by Viish** [comment]
That's not the issue:
```
2020-07-22 09:55:50:225 [liblinphone] MESSAGE Push notification received
2020-07-22 09:55:50:225 [liblinphone] ERROR AndroidPlatformHelpers::setDnsServers() exception.
```
At some point there is an exception thrown while trying to configure the DNS servers in the Core, and it never recovers from this, hence all subsequent network requests failing.

I recommend you to switch to 4.4 SDK, the issue is most probably already fixed.

**Comment by JitainSharma** [comment]
okay, let me try with 4.4 SDK will share the status.
Also, I have to support H263, H263+ in my app, does 4.4 supports it.

**Comment by Viish** [comment]
That depends from where you are getting the H263 support.

**Comment by JitainSharma** [comment]
Hi, I have used the branch **feature/kotlin** and run the debug app with background mode enable.

App works well in normal device conditions and in normal day to day doze mode as well, **but fails in long doze mode period** like in **overnight from 2020/07/22 09:00pm to morning 2020/07/23 9:42am**, when I called to the the user logged in to the device left overnight, no incoming call is received and all subsequent calls failed. After 3 calls, I wake the phone screen by tapping on the screen [Still locked] and then called again, it received the call.

Attaching the full log file uploaded to linphone: [https://www.linphone.org:444//tmp/5f1922484ca6c_021046bacb55fba7bf30.gz]

Please help to resolve this only particular scenario issue, other than that all is working perfectly fine. This is the most important use case of our app and for our users as well.

Thanks for proactively listening and helping out.

**Comment by Viish** [comment]
Hi,
Sadly in your logs there are only "error" logs and no verbose logs from that time, so it doesn't help much.
I see this is a Samsung device. In the app settings (not Linphone but the app settings from the app list in the device settings), what is the battery mode? High performance, Optimized, Medium power saving or Maximum power saving?

**Comment by JitainSharma** [comment]
Yes it's a Samsung Galaxy M21 device. Battery power mode is Medium power saving. Attaching the screenshot as well.

![Screenshot_20200723-125512_Device care](https://user-images.githubusercontent.com/2665282/88264026-af13a280-cce8-11ea-9385-62e12c64a394.jpg)
![Screenshot_20200723-133101_Settings](https://user-images.githubusercontent.com/2665282/88264260-19c4de00-cce9-11ea-9958-98e15e4d50c4.jpg)
![Screenshot_20200723-133141_Settings](https://user-images.githubusercontent.com/2665282/88264269-1df0fb80-cce9-11ea-9f67-330adfc74e37.jpg)
![Screenshot_20200723-133152_Settings](https://user-images.githubusercontent.com/2665282/88264272-1e899200-cce9-11ea-85d1-1db1c96ff388.jpg)

**Comment by Viish** [comment]
Do you reproduce the issue if you switch the battery mode to "Optimized"?

**Comment by JitainSharma** [comment]
Let me test with this mode as well.

**Comment by JitainSharma** [comment]
Hi, I have tested with the Optimized battery option and found the first two call is missed but when dial third call is got picked by the phone.

Attaching the logs file [https://www.linphone.org:444//tmp/5f1971de0f9cb_56b7f416fbff2446f03f.gz](url)

**Comment by Viish** [comment]
Can you try with SDK 4.5.0-alpha.35+d0a10f9 instead of 4.5.0-alpha.41+f536df1 please? I believe this might be caused by a change in the liblinphone project causing a regression.

**Comment by JitainSharma** [comment]
okay let me try this as well.

**Comment by JitainSharma** [comment]
Hi, I have run with the SDK 4.5.0-alpha.35+d0a10f9 version and found following observations:

*Build started from 2020/7/23 08:20pm to 2020/7/24 09:44am with Optimized battery option
1. Incoming call received at 11:09pm after 2hrs 30mins in ideal mode.
2. Incoming call failed to receive after long ideal mode between 11:10pm to 09:40am next day morning. Call time is 09:40am, 09:41am and fails until I wake the phone by tapping on it.

I removed the app from battery optimization option in the morning and tried calling it doesn't failed since the morning.

Attaching the logs file: [https://www.linphone.org:444//tmp/5f1a9f3548299_0d30b2835aae48ed886d.gz](url)

**Comment by Viish** [comment]
Hi,
We have found an infinite loop issue on master branch. Fix is in 4.5.0-alpha-48-g2ea8a2f and newer.
I think this will solve the last issues you had.
Cheers,
